text/0007-gas-price-adjustment.md

```diff
@@ -0,0 +1,135 @@
+
+The update can be either a percent increase or decrease of `BASE_PRICE`, specified by an `ADJUSTMENT_RATE`. To decide on the direction of the change, we make use of the fullnesses of blocks from the preceding adjustment period. We map the array of fullnesses to a single value called `FULLNESS` (details on how to compute this are given below). Then, if `FULLNESS` is greater than `TARGET_FULLNESS`, we increase the price. Otherwise, we decrease the price.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+### Protocol Variables
+
+- `BASE_PRICE`: The value of the price floor, in motes. Given blocks proposed in a certain adjustment period, all transactions must be charged at least this value. The value will be included in each block in a field of the same name.
+- `FULLNESS`: A value representing the fullness of blocks of a given adjustment period.
+
+### Protocol Parameters
+
+- `INITIAL_BASE_PRICE`: The value of `BASE_PRICE` at genesis. Also the *minimum value* for `BASE_PRICE`. Value: TBD.
+- `ADJUSTMENT_RATE`: The rate at which `BASE_PRICE` can move up or down per day. Value: `10_000_000_000` (1%).
+- `DENOMINATOR`: Fractions will be represented as trillionths of 1. Value: `1_000_000_000_000`.
+- `BLOCK_GAS_LIMIT`: Maximum gas allowed in a block. Value: TBD.
+- `TARGET_FULLNESS`: The value which will determine whether `BASE_PRICE` should move up or down. Value: `650_000_000_000` (65%).
+- `MIN_FINALIZED_BLOCKS`: The minimum number of blocks that must be finalized in an adjustment period. Otherwise, a fail-safe is triggered. Value: TBD.
+
+### Transaction Fields
+
+- `PRICE_PREMIUM`: The price premium specified for a given transaction.
+- `MAX_GAS_PRICE`: The maximum gas price which the paying signer is willing to accept. The maximum `BASE_PRICE` they are willing to accept is computed as `MAX_GAS_PRICE - PRICE_PREMIUM`.
+
+### Transaction Variables
+
+- `GAS_USED`: The total gas used by a given transaction.
+
+### Adjustment Process
+
+For the `n`-th adjustment period, all finalized blocks with a timestamp in the range `[n*ADJUSTMENT_PERIOD, (n+1)*ADJUSTMENT_PERIOD]` are considered. The total amount of gas used in each one is collected in an array `gas_used_array`. `FULLNESS` for the `n`-th period is computed as:
+
+```python
+FULLNESS = DENOMINATOR * median(gas_used_array) / BLOCK_GAS_LIMIT
+```
+
+Then, a system transaction can be executed in any block proposed after `(n+1)*ADJUSTMENT_PERIOD`, which updates `BASE_PRICE` according to:
+
+```python
+if FULLNESS > TARGET_FULLNESS:
+    new_base_price = BASE_PRICE * (DENOMINATOR + ADJUSTMENT_RATE) / DENOMINATOR
+else:
+    new_base_price = BASE_PRICE * DENOMINATOR / (DENOMINATOR + ADJUSTMENT_RATE)
+
+if new_base_price > INITIAL_BASE_PRICE:
+    BASE_PRICE = new_base_price
```

Comment:
Or maybe:
```python
BASE_PRICE = max(new_base_price, INITIAL_BASE_PRICE)
```
Because if it's lower, then at least we can go back to `INITIAL_BASE_PRICE`?

Reply:
`BASE_PRICE` should not be able to go lower than `INITIAL_BASE_PRICE`. Did I make a mistake with the pseudocode?

Reply:
No, but the pseudocode would leave the `BASE_PRICE` at `105` if `INITIAL_BASE_PRICE` is `100` and `new_base_price` is `90`. I think setting it to `100` would make more sense.

Also, it's a strict inequality, so even if `new_base_price` is `100` and `INITIAL_BASE_PRICE` is `100`, it would leave `BASE_PRICE` at `105` instead of setting it to `100`.

Reply:
I realize it now, sorry. Fixed